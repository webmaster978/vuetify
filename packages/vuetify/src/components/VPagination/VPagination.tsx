import './VPagination.sass'

// Types
import type { Density } from '@/composables/density'

// Components
import { VBtn } from '../VBtn'

// Utilities
import { ComponentPublicInstance, computed, defineComponent, nextTick, ref } from 'vue'
import { createRange, keyCodes, makeProps } from '@/util'

// Composables
import { makeTagProps } from '@/composables/tag'
import { useLocale } from '@/composables/locale'
import { useRtl } from '@/composables/rtl'
import { makeElevationProps } from '@/composables/elevation'
import { makeDensityProps } from '@/composables/density'
import { makeRoundedProps } from '@/composables/rounded'
import { makeSizeProps } from '@/composables/size'
import { useResizeObserver } from '@/composables/resizeObserver'
import { makeBorderProps } from '@/composables/border'
import { useRefs } from '@/composables/refs'
import { useProxiedModel } from '@/composables/proxiedModel'

export default defineComponent({
  name: 'VPagination',

  props: makeProps({
    modelValue: {
      type: Number,
      default: 1,
    },
    disabled: Boolean,
    length: {
      type: Number,
      default: 0,
      validator: (val: number) => val % 1 === 0,
    },
    start: {
      type: Number,
      default: 1,
    },
    totalVisible: [Number, String],
    nextIcon: {
      type: String,
      default: '$next',
    },
    prevIcon: {
      type: String,
      default: '$prev',
    },
    firstIcon: {
      type: String,
      default: '$first',
    },
    lastIcon: {
      type: String,
      default: '$last',
    },
    pageAriaLabel: {
      type: String,
      default: '$vuetify.pagination.ariaLabel.page',
    },
    currentPageAriaLabel: {
      type: String,
      default: '$vuetify.pagination.ariaLabel.currentPage',
    },
    previousAriaLabel: {
      type: String,
      default: '$vuetify.pagination.ariaLabel.previous',
    },
    nextAriaLabel: {
      type: String,
      default: '$vuetify.pagination.ariaLabel.next',
    },
    ariaLabel: {
      type: String,
      default: '$vuetify.pagination.ariaLabel.root',
    },
    firstAriaLabel: {
      type: String,
      default: '$vuetify.pagination.ariaLabel.first',
    },
    lastAriaLabel: {
      type: String,
      default: '$vuetify.pagination.ariaLabel.last',
    },
    color: {
      type: [String, Boolean],
      default: 'primary',
    },
    showFirstLastPage: Boolean,
    ...makeTagProps({ tag: 'nav' }),
    ...makeElevationProps({ elevation: 1 }),
    ...makeDensityProps({ density: 'comfortable' as Density }),
    ...makeRoundedProps({ rounded: 'sm' }),
    ...makeSizeProps(),
    ...makeBorderProps(),
  }),

  setup (props, ctx) {
    const page = useProxiedModel(props, 'modelValue')
    const { t } = useLocale()
    const { isRtl } = useRtl()
    const { resizeRef } = useResizeObserver(entries => {
      if (!entries.length) return

      const { target, contentRect } = entries[0]

      const firstItem = target.querySelector('.v-pagination__list > *')

      if (!firstItem) return

      const totalWidth = contentRect.width
      const itemWidth = firstItem.getBoundingClientRect().width + 10

      maxButtons.value = Math.max(0, Math.floor((totalWidth - 96)/ itemWidth))
    })

    const maxButtons = ref(-1)

    const totalVisible = computed(() => {
      if (props.totalVisible) return Math.min(parseInt(props.totalVisible ?? '', 10), props.length)
      else if (maxButtons.value >= 0) return maxButtons.value
      return props.length
    })

    const range = computed(() => {
      if (props.length <= 0) return []

      if (totalVisible.value <= 3) {
        return [Math.min(Math.max(props.start, page.value), props.start + props.length)]
      }

      if (props.length <= totalVisible.value) {
        return createRange(props.length, props.start)
      }

      const even = totalVisible.value % 2 === 0 ? 1 : 0
      const middle = Math.ceil(totalVisible.value / 2)
      const left = middle + even
      const right = props.length - middle + even

      if (page.value < left) {
        return [...createRange(Math.max(1, totalVisible.value - 2), props.start), '...', props.length]
      } else if (page.value > right) {
        const length = totalVisible.value - 2
        const start = props.length - length + props.start
        return [props.start, '...', ...createRange(length, start)]
      } else {
        const length = Math.max(1, totalVisible.value - 4)
        const start = page.value - Math.floor(length / 2) + props.start
        return [props.start, '...', ...createRange(length, start), '...', props.length]
      }
    })

    function emit (e: Event, value: number, event?: string) {
      e.preventDefault()
      page.value = value
      event && ctx.emit(event, value)
    }

    const { refs, updateRef } = useRefs<ComponentPublicInstance>()

    const items = computed(() => {
      const sharedProps = {
        density: props.density,
        rounded: props.rounded,
        size: props.size,
      }

      return range.value.map((item, index) => {
        if (typeof item === 'string') {
          return {
            ...sharedProps,
            ref: (e: any) => updateRef(e, index),
            page: item,
            icon: true,
            disabled: true,
            text: true,
            outlined: props.outlined,
            border: props.border,
          }
        } else {
          return {
            ...sharedProps,
            ref: (e: any) => updateRef(e, index),
            page: item,
            icon: true,
            disabled: !!props.disabled,
            elevation: props.elevation,
            outlined: props.outlined,
            border: props.border,
            text: item !== page.value,
            color: item === page.value ? props.color : false,
            ariaCurrent: item === page.value,
            ariaLabel: t(item === page.value ? props.currentPageAriaLabel : props.pageAriaLabel, index + 1),
            onClick: (e: Event) => emit(e, item),
          }
        }
      })
    })

    const controls = computed(() => {
      const sharedProps = {
        color: false,
        density: props.density,
        rounded: props.rounded,
        size: props.size,
        text: true,
        outlined: props.outlined,
        border: props.border,
      }

      return {
        first: props.showFirstLastPage ? {
          ...sharedProps,
          icon: isRtl.value ? props.lastIcon : props.firstIcon,
          onClick: (e: Event) => emit(e, props.start, 'first'),
          disabled: !!props.disabled || page.value <= props.start,
          ariaLabel: t(props.firstAriaLabel),
        } : undefined,
        prev: {
          ...sharedProps,
          icon: isRtl.value ? props.nextIcon : props.prevIcon,
          onClick: (e: Event) => emit(e, page.value - 1, 'prev'),
          disabled: !!props.disabled || page.value <= props.start,
          ariaLabel: t(props.previousAriaLabel),
        },
        next: {
          ...sharedProps,
          icon: isRtl.value ? props.prevIcon : props.nextIcon,
          onClick: (e: Event) => emit(e, page.value + 1, 'next'),
          disabled: !!props.disabled || page.value >= props.start + props.length - 1,
          ariaLabel: t(props.nextAriaLabel),
        },
        last: props.showFirstLastPage ? {
          ...sharedProps,
          icon: isRtl.value ? props.firstIcon : props.lastIcon,
          onClick: (e: Event) => emit(e, props.start + props.length - 1, 'last'),
          disabled: !!props.disabled || page.value >= props.start + props.length - 1,
          ariaLabel: t(props.lastAriaLabel),
        } : undefined,
      }
    })

    function updateFocus () {
      const currentIndex = page.value - props.start
      refs.value[currentIndex]?.$el.focus()
    }

    function onKeydown (e: KeyboardEvent) {
      if (e.keyCode === keyCodes.left && !props.disabled && page.value > props.start) {
        page.value = page.value - 1
        nextTick(updateFocus)

      } else if (e.keyCode === keyCodes.right && !props.disabled && page.value < props.start + props.length - 1) {
        page.value = page.value + 1
        nextTick(updateFocus)
      }
    }

    return () => (
      <props.tag
        ref={resizeRef}
        class='v-pagination'
        role='navigation'
        aria-label={t(props.ariaLabel)}
        onKeydown={onKeydown}
      >
        <ul class='v-pagination__list'>
          {props.showFirstLastPage && (
            <li class='v-pagination__first'>
              { ctx.slots.first ? ctx.slots.first(controls.value.first) : (
                <VBtn {...controls.value.first} />
              )}
            </li>
          )}

          <li class='v-pagination__prev'>
            { ctx.slots.prev ? ctx.slots.prev(controls.value.prev) : (
              <VBtn {...controls.value.prev} />
            )}
          </li>

          { items.value.map(({ page, ...item }, index) => (
            <li
              key={`${index}_${page}`}
              class='v-pagination__item'
            >
              { ctx.slots.item ? ctx.slots.item({ page, ...item }) : (
                <VBtn {...item}>{page}</VBtn>
              ) }
            </li>
          )) }

          <li class='v-pagination__next'>
            { ctx.slots.next ? ctx.slots.next(controls.value.next) : (
              <VBtn {...controls.value.next} />
            )}
          </li>

          {props.showFirstLastPage && (
            <li class='v-pagination__last'>
              { ctx.slots.last ? ctx.slots.last(controls.value.last) : (
                <VBtn {...controls.value.last} />
              )}
            </li>
          )}
        </ul>
      </props.tag>
    )
  }
})
