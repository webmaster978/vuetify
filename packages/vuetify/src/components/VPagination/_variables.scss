@import '../../styles/styles.sass';

$pagination-item-margin: .3rem !default;

