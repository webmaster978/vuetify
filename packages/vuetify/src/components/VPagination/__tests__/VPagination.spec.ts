import VPagination from '../VPagination'

import { mount } from '@vue/test-utils'
import { createVuetify } from '@/framework'

describe('VPagination.ts', () => {
  const vuetify = createVuetify()
  const mountFunction = (options?: any) => mount(VPagination, {
    ...options,
    global: {
      plugins: [vuetify],
    },
  })

  it('emits an event when next or previous is clicked', async () => {
    const wrapper = mountFunction({
      props: {
        length: 5,
        modelValue: 2,
      },
    })

    const prev = wrapper.find('.v-pagination__prev > .v-btn')
    await prev.trigger('click')

    expect(wrapper.emitted('prev')).toEqual([[1]])

    const next = wrapper.find('.v-pagination__next > .v-btn')
    await next.trigger('click')

    expect(wrapper.emitted('next')).toEqual([[3]])
  })

  it.skip('should render component in RTL mode and match snapshot', async () => {
    const wrapper = mountFunction({
      props: {
        length: 5,
        modelValue: 2,
      },
    })

    wrapper.vm.$vuetify.rtl = true
    await wrapper.vm.$nextTick()

    expect(wrapper.html()).toMatchSnapshot()
    wrapper.vm.$vuetify.rtl = undefined
  })

  it('emits an event when pagination item is clicked', async () => {
    const wrapper = mountFunction({
      props: {
        length: 5,
        modelValue: 1,
      },
    })

    const navigation = wrapper.findAll('.v-pagination__item > .v-btn')
    navigation[1].trigger('click')

    expect(wrapper.emitted('update:modelValue')).toEqual([[2]])
  })

  it('should render disabled buttons with length equals to 0', async () => {
    const wrapper = mountFunction({
      props: {
        length: 0,
        modelValue: 1,
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should use totalVisible prop if defined', async () => {
    const wrapper = mountFunction({
      props: {
        length: 100,
        modelValue: 50,
        totalVisible: 10,
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render correct pages when using start prop', async () => {
    const wrapper = mountFunction({
      props: {
        start: 3,
        length: 10,
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  // https://github.com/vuetifyjs/vuetify/issues/7947
  it.skip('should never show more than the max number of allowed buttons', () => {
    const wrapper = mountFunction({
      data: () => ({
        maxButtons: 4,
      }),

      props: {
        length: 40,
        totalVisible: 10,
      },
    })

    expect(wrapper.vm.items).toHaveLength(4)

    wrapper.setData({ maxButtons: 12 })

    expect(wrapper.vm.items).toHaveLength(10)
  })

  it.skip('should never show more than the number of total visible buttons', () => {
    const wrapper = mountFunction({
      data: () => ({
        maxButtons: 0,
      }),

      props: {
        length: 5,
        totalVisible: undefined,
      },
    })

    expect(wrapper.vm.items).toHaveLength(5)

    wrapper.setProps({ length: 40 })

    wrapper.setData({ maxButtons: 0 })
    wrapper.setProps({ totalVisible: 10 })
    expect(wrapper.vm.items).toHaveLength(10)

    wrapper.setData({ maxButtons: 11 })
    wrapper.setProps({ totalVisible: undefined })
    expect(wrapper.vm.items).toHaveLength(11)

    wrapper.setData({ maxButtons: 12 })
    wrapper.setProps({ totalVisible: 13 })
    expect(wrapper.vm.items).toHaveLength(12)
  })

  it.skip('should return length when maxButtons is less than 1', () => {
    const wrapper = mountFunction({
      data: () => ({ maxButtons: -3 }),
      props: { length: 4 },
    })

    expect(wrapper.vm.items).toEqual([1, 2, 3, 4])
  })
})
