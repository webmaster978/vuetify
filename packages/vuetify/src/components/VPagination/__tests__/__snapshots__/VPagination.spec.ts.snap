// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VPagination.ts should render component in RTL mode and match snapshot 1`] = `
<nav role="navigation"
     aria-label="$vuetify.pagination.ariaLabel.wrapper"
>
  <ul class="v-pagination theme--light">
    <li>
      <button type="button"
              aria-label="$vuetify.pagination.ariaLabel.previous"
              class="v-pagination__navigation"
      >
        <i aria-hidden="true"
           class="v-icon notranslate mdi mdi-chevron-right theme--light"
        >
        </i>
      </button>
    </li>
    <li>
      <button type="button"
              aria-label="$vuetify.pagination.ariaLabel.page"
              class="v-pagination__item"
      >
        1
      </button>
    </li>
    <li>
      <button type="button"
              aria-current="true"
              aria-label="$vuetify.pagination.ariaLabel.currentPage"
              class="v-pagination__item v-pagination__item--active primary"
      >
        2
      </button>
    </li>
    <li>
      <button type="button"
              aria-label="$vuetify.pagination.ariaLabel.page"
              class="v-pagination__item"
      >
        3
      </button>
    </li>
    <li>
      <button type="button"
              aria-label="$vuetify.pagination.ariaLabel.page"
              class="v-pagination__item"
      >
        4
      </button>
    </li>
    <li>
      <button type="button"
              aria-label="$vuetify.pagination.ariaLabel.page"
              class="v-pagination__item"
      >
        5
      </button>
    </li>
    <li>
      <button type="button"
              aria-label="$vuetify.pagination.ariaLabel.next"
              class="v-pagination__navigation"
      >
        <i aria-hidden="true"
           class="v-icon notranslate mdi mdi-chevron-left theme--light"
        >
        </i>
      </button>
    </li>
  </ul>
</nav>
`;

exports[`VPagination.ts should render correct pages when using start prop 1`] = `
<nav class="v-pagination"
     role="navigation"
     aria-label="Pagination Navigation"
>
  <ul class="v-pagination__list">
    <li class="v-pagination__prev">
      <button type="button"
              class="v-btn v-btn--icon v-btn--disabled v-theme--light v-btn--density-comfortable rounded-sm v-btn--size-default"
              disabled
              arialabel="Previous page"
      >
        <span class="v-btn__overlay">
        </span>
        <i class="mdi-chevron-left mdi v-icon notranslate v-icon--size-default"
           aria-hidden="true"
        >
        </i>
      </button>
    </li>
    <li class="v-pagination__item">
      <button type="button"
              class="v-btn v-btn--icon v-theme--light v-btn--border v-btn--density-comfortable elevation-1 rounded-sm v-btn--size-default"
              ariacurrent="false"
              arialabel="Goto Page 1"
      >
        <span class="v-btn__overlay">
        </span>
        3
      </button>
    </li>
    <li class="v-pagination__item">
      <button type="button"
              class="v-btn v-btn--icon v-theme--light v-btn--border v-btn--density-comfortable elevation-1 rounded-sm v-btn--size-default"
              ariacurrent="false"
              arialabel="Goto Page 2"
      >
        <span class="v-btn__overlay">
        </span>
        4
      </button>
    </li>
    <li class="v-pagination__item">
      <button type="button"
              class="v-btn v-btn--icon v-theme--light v-btn--border v-btn--density-comfortable elevation-1 rounded-sm v-btn--size-default"
              ariacurrent="false"
              arialabel="Goto Page 3"
      >
        <span class="v-btn__overlay">
        </span>
        5
      </button>
    </li>
    <li class="v-pagination__item">
      <button type="button"
              class="v-btn v-btn--icon v-theme--light v-btn--border v-btn--density-comfortable elevation-1 rounded-sm v-btn--size-default"
              ariacurrent="false"
              arialabel="Goto Page 4"
      >
        <span class="v-btn__overlay">
        </span>
        6
      </button>
    </li>
    <li class="v-pagination__item">
      <button type="button"
              class="v-btn v-btn--icon v-theme--light v-btn--border v-btn--density-comfortable elevation-1 rounded-sm v-btn--size-default"
              ariacurrent="false"
              arialabel="Goto Page 5"
      >
        <span class="v-btn__overlay">
        </span>
        7
      </button>
    </li>
    <li class="v-pagination__item">
      <button type="button"
              class="v-btn v-btn--icon v-theme--light v-btn--border v-btn--density-comfortable elevation-1 rounded-sm v-btn--size-default"
              ariacurrent="false"
              arialabel="Goto Page 6"
      >
        <span class="v-btn__overlay">
        </span>
        8
      </button>
    </li>
    <li class="v-pagination__item">
      <button type="button"
              class="v-btn v-btn--icon v-theme--light v-btn--border v-btn--density-comfortable elevation-1 rounded-sm v-btn--size-default"
              ariacurrent="false"
              arialabel="Goto Page 7"
      >
        <span class="v-btn__overlay">
        </span>
        9
      </button>
    </li>
    <li class="v-pagination__item">
      <button type="button"
              class="v-btn v-btn--icon v-theme--light v-btn--border v-btn--density-comfortable elevation-1 rounded-sm v-btn--size-default"
              ariacurrent="false"
              arialabel="Goto Page 8"
      >
        <span class="v-btn__overlay">
        </span>
        10
      </button>
    </li>
    <li class="v-pagination__item">
      <button type="button"
              class="v-btn v-btn--icon v-theme--light v-btn--border v-btn--density-comfortable elevation-1 rounded-sm v-btn--size-default"
              ariacurrent="false"
              arialabel="Goto Page 9"
      >
        <span class="v-btn__overlay">
        </span>
        11
      </button>
    </li>
    <li class="v-pagination__item">
      <button type="button"
              class="v-btn v-btn--icon v-theme--light v-btn--border v-btn--density-comfortable elevation-1 rounded-sm v-btn--size-default"
              ariacurrent="false"
              arialabel="Goto Page 10"
      >
        <span class="v-btn__overlay">
        </span>
        12
      </button>
    </li>
    <li class="v-pagination__next">
      <button type="button"
              class="v-btn v-btn--icon v-theme--light v-btn--density-comfortable rounded-sm v-btn--size-default"
              arialabel="Next page"
      >
        <span class="v-btn__overlay">
        </span>
        <i class="mdi-chevron-right mdi v-icon notranslate v-icon--size-default"
           aria-hidden="true"
        >
        </i>
      </button>
    </li>
  </ul>
</nav>
`;

exports[`VPagination.ts should render disabled buttons with length equals to 0 1`] = `
<nav class="v-pagination"
     role="navigation"
     aria-label="Pagination Navigation"
>
  <ul class="v-pagination__list">
    <li class="v-pagination__prev">
      <button type="button"
              class="v-btn v-btn--icon v-btn--disabled v-theme--light v-btn--density-comfortable rounded-sm v-btn--size-default"
              disabled
              arialabel="Previous page"
      >
        <span class="v-btn__overlay">
        </span>
        <i class="mdi-chevron-left mdi v-icon notranslate v-icon--size-default"
           aria-hidden="true"
        >
        </i>
      </button>
    </li>
    <li class="v-pagination__next">
      <button type="button"
              class="v-btn v-btn--icon v-btn--disabled v-theme--light v-btn--density-comfortable rounded-sm v-btn--size-default"
              disabled
              arialabel="Next page"
      >
        <span class="v-btn__overlay">
        </span>
        <i class="mdi-chevron-right mdi v-icon notranslate v-icon--size-default"
           aria-hidden="true"
        >
        </i>
      </button>
    </li>
  </ul>
</nav>
`;

exports[`VPagination.ts should use totalVisible prop if defined 1`] = `
<nav class="v-pagination"
     role="navigation"
     aria-label="Pagination Navigation"
>
  <ul class="v-pagination__list">
    <li class="v-pagination__prev">
      <button type="button"
              class="v-btn v-btn--icon v-theme--light v-btn--density-comfortable rounded-sm v-btn--size-default"
              arialabel="Previous page"
      >
        <span class="v-btn__overlay">
        </span>
        <i class="mdi-chevron-left mdi v-icon notranslate v-icon--size-default"
           aria-hidden="true"
        >
        </i>
      </button>
    </li>
    <li class="v-pagination__item">
      <button type="button"
              class="v-btn v-btn--icon v-theme--light v-btn--border v-btn--density-comfortable elevation-1 rounded-sm v-btn--size-default"
              ariacurrent="false"
              arialabel="Goto Page 1"
      >
        <span class="v-btn__overlay">
        </span>
        1
      </button>
    </li>
    <li class="v-pagination__item">
      <button type="button"
              class="v-btn v-btn--icon v-btn--disabled v-theme--light text-primary v-btn--density-comfortable rounded-sm v-btn--size-default"
              disabled
      >
        <span class="v-btn__overlay">
        </span>
        ...
      </button>
    </li>
    <li class="v-pagination__item">
      <button type="button"
              class="v-btn v-btn--icon v-theme--light v-btn--border v-btn--density-comfortable elevation-1 rounded-sm v-btn--size-default"
              ariacurrent="false"
              arialabel="Goto Page 3"
      >
        <span class="v-btn__overlay">
        </span>
        48
      </button>
    </li>
    <li class="v-pagination__item">
      <button type="button"
              class="v-btn v-btn--icon v-theme--light v-btn--border v-btn--density-comfortable elevation-1 rounded-sm v-btn--size-default"
              ariacurrent="false"
              arialabel="Goto Page 4"
      >
        <span class="v-btn__overlay">
        </span>
        49
      </button>
    </li>
    <li class="v-pagination__item">
      <button type="button"
              class="v-btn v-btn--contained v-btn--elevated v-btn--icon v-theme--light bg-primary v-btn--density-comfortable elevation-1 rounded-sm v-btn--size-default"
              ariacurrent="true"
              arialabel="Page 5, Current Page"
      >
        <span class="v-btn__overlay">
        </span>
        50
      </button>
    </li>
    <li class="v-pagination__item">
      <button type="button"
              class="v-btn v-btn--icon v-theme--light v-btn--border v-btn--density-comfortable elevation-1 rounded-sm v-btn--size-default"
              ariacurrent="false"
              arialabel="Goto Page 6"
      >
        <span class="v-btn__overlay">
        </span>
        51
      </button>
    </li>
    <li class="v-pagination__item">
      <button type="button"
              class="v-btn v-btn--icon v-theme--light v-btn--border v-btn--density-comfortable elevation-1 rounded-sm v-btn--size-default"
              ariacurrent="false"
              arialabel="Goto Page 7"
      >
        <span class="v-btn__overlay">
        </span>
        52
      </button>
    </li>
    <li class="v-pagination__item">
      <button type="button"
              class="v-btn v-btn--icon v-theme--light v-btn--border v-btn--density-comfortable elevation-1 rounded-sm v-btn--size-default"
              ariacurrent="false"
              arialabel="Goto Page 8"
      >
        <span class="v-btn__overlay">
        </span>
        53
      </button>
    </li>
    <li class="v-pagination__item">
      <button type="button"
              class="v-btn v-btn--icon v-btn--disabled v-theme--light text-primary v-btn--density-comfortable rounded-sm v-btn--size-default"
              disabled
      >
        <span class="v-btn__overlay">
        </span>
        ...
      </button>
    </li>
    <li class="v-pagination__item">
      <button type="button"
              class="v-btn v-btn--icon v-theme--light v-btn--border v-btn--density-comfortable elevation-1 rounded-sm v-btn--size-default"
              ariacurrent="false"
              arialabel="Goto Page 10"
      >
        <span class="v-btn__overlay">
        </span>
        100
      </button>
    </li>
    <li class="v-pagination__next">
      <button type="button"
              class="v-btn v-btn--icon v-theme--light v-btn--density-comfortable rounded-sm v-btn--size-default"
              arialabel="Next page"
      >
        <span class="v-btn__overlay">
        </span>
        <i class="mdi-chevron-right mdi v-icon notranslate v-icon--size-default"
           aria-hidden="true"
        >
        </i>
      </button>
    </li>
  </ul>
</nav>
`;
